### ISSUE TITLE
08_path1a3 test problem

### ISSUE THREAD

**Issue by PhilterPaper** [sciurius/perl-SVGPDF#8]
I don't know exactly what problem you encountered with this test, but it mentions failing on an extended precision library (I presume in the CPANTS results). I saw this problem with PDF::Table, where direct comparisons of "raw" floating point numbers would fail because test systems built with different extended precision libraries would produce slightly different results (starting at perhaps the 20th significant digit). The solution I came up with was to _round_ the raw floating point numbers to a much smaller number of significant digits (6 to 12), which are the same no matter what the library or hardware used is, and compare against that short result. So far, it's worked well, but sometimes needed several iterations of tests to catch all the cases (i.e., a test now failing later, and showing a different floating point result that's failing). More significant digits than that are seldom of importance in this kind of work.

**Comment by sciurius** [comment]
If you have a system that has this test failing in 0.087 maybe you can take a look. It is the call to 
````
bogen(80.00,230.00 125.00,275.00 45.00 move=0 large=0 dir=1 rot=0.00)
````
This expands to 4 curves:
````
      curve(  85.91,230.00  91.76,231.16  97.22,233.43 )
      curve( 102.68,235.69 107.64,239.00 111.82,243.18 )
      curve( 116.00,247.36 119.31,252.32 121.57,257.78 )
      curve( 123.84,263.24 125.00,269.09 125.00,275.00 )
````
In a system that fails, there's a 5th curve:
````
      curve( 125.00,275.00 125.00,275.00 125.00,275.00 )
````
This is an extremely small curve that should not be there. Apparently the calculation misses the endpoint due to math precision.

**Comment by PhilterPaper** [comment]
No, I do not have failures installing 0.087. It sounds like the "tiny last B-spline" problem you found in PhilterPaper/Perl-PDF-Builder/issues/208. 

I thought it might be the t-test extended-precision floating point problem I encountered in some of my libraries, but now it sounds more like the bogen precision issue you raised. I really _have_ to get 3.027 out the door as my top priority, but it sounds like I need to fix this issue in PDF::Builder and SVGPDF bogen code as a very high priority right after that.

**Comment by PhilterPaper** [comment]
I'll leave this open for now, to remind me that I owe you a patch for the bogen/elliptical arc code I gave you for SVGPDF. If you make your own fix, please close and let me know that there's some code for me to look at.

**Comment by PhilterPaper** [comment]
Given your interest in tree rats :-) you might find interesting https://www.youtube.com/watch?v=_tW9pfCqIgY between 20:45 and 22:35.

**Comment by PhilterPaper** [comment]
I think I cured the problem in PDF::Builder \#208. It affects only `bogen()` (not `arc()`, `pie()`, `curve()`, or anyone calling them), and is caused by the two sweep angles calculation creating a very tiny "noise" remnant (around 1e-14 degree) that causes a sweep across the 0 degree line, in turn causing a very tiny segment. This fix forces that tiny angle to exactly 0.

The code segment
```
    # always draw CW (dir=1)
    # note that start and end could be well out of +/-360 degree range
    @points = _arctocurve($r,$r, 90+$alpha+$beta/2,90+$alpha-$beta/2, 1);
```
is replaced by
```
    # always draw CW (dir=1)
    # note that start and end could be well out of +/-360 degree range
    my $sweep1 = 90 + $alpha + $beta/2;
    my $sweep2 = 90 + $alpha - $beta/2;
    if (abs($sweep1*$r) < 0.001) {
        # tiny angle force to 0
        $sweep1 = 0;
    }
    @points = _arctocurve($r,$r, $sweep1,$sweep2, 1);
```
I think your bogen.pm code can take this as-is for the bogen() code. I didn't do a PR because I'm not sure if any of the other elliptical etc. routines in bogen.pm are affected. If you're using two different radii ($rx and $ry), you should probably use `$sweep*max($rx,$ry)` to cover and test the largest segment.

Please try running with the revised code, and see if it cures your issue with unnecessary curve calls. If not, we will need to dig deeper.

**Comment by sciurius** [comment]
Thanks, Phil. SGVPDF 0.69 is on its way to CPAN and we'll soon hear the results of the testers.

**Comment by PhilterPaper** [comment]
SVGPDF 0.92 (not 0.69) installs successfully on Windows (Strawberry Perl) 5.28 through 5.42. However, the Testers Matrix shows a number of 08_path1a3 and 08_path1a3-ld test failures on Linux and BSD. There may be other failures too (the two Tester sections seem a bit different).

I see Github hasn't been updated yet (lib/, t/, xt/), just in case you overlooked that...

Is SVGPDF to the stage where you can call it production? Version 1.00 and "**WARNING** This is a work in progress and probably not useful yet." removed?

**Comment by sciurius** [comment]
There were separate tests for longdouble perls, and these backfired now the underlying issue (bogen) was fixed. I've released 0.92.1 to fix this (by removing the special-case).

**Comment by sciurius** [comment]
> Is SVGPDF to the stage where you can call it production? Version 1.00 and "WARNING This is a work in progress and probably not useful yet." removed?

Given that there are so many SVG and CSS features not supported, I would not call it 'production quality'. Even though it does its job for you and me.

**Comment by PhilterPaper** [comment]
> Given that there are so many SVG and CSS features not supported, I would not call it 'production quality'. Even though it does its job for you and me.

I'm just thinking that you may be scaring off potential users by telling them it's not a useful library for them to try using. It also implies that a lot of functionality is still in the process of being actively added, and if a potential user just waits a short time, it may be adequate for them. How about instead:

> **NOTICE:** While providing useful support for basic SVG functionality, this library is far from a complete implementation of SVG or CSS. Many SVG files found _in the wild_ will work, but others will find it lacks needed features. Use at your own risk, understanding what SVG features you need and what SVGPDF provides.

**Comment by sciurius** [comment]
Right. However this would imply that there is a planned and steady (but maybe slow) development going on to eventually add the missing featues — which is not the case. At least not as far as I am concerned.

I developed SVGPDF for my own specific purposes, and added a number of features on your request, mostly since I considered it an interesting challenge, and you contributed code and useful testing materials.

But my time and energy is limited, and merely re-inventing existing software is not on the top of my list. I anticipated (hoped) that after a while it would be possible to directly use the real implementation libraries that browsers use but there seem to be little or no progress in that direction. Also, I think a smart guy with AI assistance can develop something like this in a fraction of the time it would take you or me.

I have no problems to raise the version to nonzero and add the disclaimer you suggest and will do so when 0.96.1 results are satisfactory.

**Comment by PhilterPaper** [comment]
> This is a work in progress and probably not useful **yet**.

I'm not sure if you were referring to the original text (above) or to what I said about it implying that additional work was being done on it (i.e., being actively developed). I fully understand that you do not intend to do a lot more work on SVGPDF, and if I need more features I will try to do the coding (maybe a PR). You just want to make it clear to users that what they see now is probably close to the final level of development this library will see. It _is_ already quite useful, so long as one understands what it does -- and does not -- support. Maybe that could be better documented? BTW, your work on this is much appreciated.

> Also, I think a smart guy with AI assistance can develop something like this in a fraction of the time it would take you or me.

I don't trust AI "developed" software. As with writing your school report in a natural language, all they do is steal bits and pieces of existing code (no matter what the quality) and slap them together, often resulting in massive security vulnerabilities and lots of bugs. Maybe as an aid to _finding_ existing code to steal, er, work off of, AI would be useful. It would be nice if the browser implementations were available and usable for this purpose, but even then, all you get out of it is the syntax parsing and maybe a data structure -- actually outputting good PDF code is still left up to the user!

**Comment by sciurius** [comment]
FYI: https://www.cpantesters.org/cpan/report/f06da8a8-3c8f-11f1-8bf1-8251589ab4ce

**Comment by PhilterPaper** [comment]
I was just looking at the Testers Matrix. Andreas appears to be using a number of pre-3.027 releases of PDF::Builder, where oddball release numbers for libraries (e.g., 0.59_01) cause failures. I think that 3.027 and up should be OK. Is there any way to contact him and ask him to update his PDF::Builder to current?

**Comment by sciurius** [comment]
The METAcpan says [andreas.koenig.7os6vvqr@franz.ak.mind.de](mailto:andreas.koenig.7os6vvqr@franz.ak.mind.de)

**Comment by PhilterPaper** [comment]
I sent a note to Andreas asking him to check whether he was running at least 3.027 on PDF::Builder. Then I saw the SVGPDF build update requiring minimum 3.027. I cc'd you on the email, but jv@cpan.org bounced (just in case you're expecting mail from people on that address!). I think their whole email system is broken -- I just tried pmperry@cpan.org and it bounced too.

Right now, of the prereqs I check the minimum version on, I think SVGPDF is the only one with an oddly formatted version number. Andreas must also have some older (but not necessarily backlevel) prereqs. I'm not sure I can (or should) do anything more about handling this. I _think_ that PDF::Builder 3.027 and up should handle them OK, but it's been so long since I put that code in that I'm not 100% positive.

**Comment by sciurius** [comment]
Hmm. Bouncing cpan.org mail is not really helping building confidence in the CPAN...
Fortunately most module-related communication goes via GitHub these days.

**Comment by PhilterPaper** [comment]
Most likely Andreas is still running a pre-3.027 PDF::Builder on a couple of his test platforms, which chokes on odd version formats, but I don't know why the new SVGPDF requirement for 3.027 would have allowed that. As no one else seems to be experiencing the problem, I'll go ahead and close this ticket (feel free to reopen it if you're not satisfied that the issue is resolved). If someone shows that PDF::Builder is _not_ handling odd version formats correctly, we can open a new ticket in PDF::Builder to deal with that. In the meantime, perhaps you could verify that SVGPDF in fact **won't** install and test with pre-3.027? It doesn't seem to be listed as a hard prereq, and thus _not_ getting a version check itself?

**Comment by sciurius** [comment]
The check in `t/901_regtest.t` looks solid, if I change my PDF::Builder version to 3.026 it bails out.

I have added the actual version number to the diag `Testing 33 SVG files with PDF::Builder` so at least we will have a visual clue.

**Comment by sciurius** [comment]
All tests for 0.092.3 are green, including Andreas' systems. However from the log it seems he upgraded PDF::Builder to 3.027.

Going to release 0.092.3 as 1.000.0.

**Comment by PhilterPaper** [comment]
Sounds good! Don't forget Readme.md disclaimer about usefulness, and copyright date (still at 2024).
